### PR TITLE
[Pallas][Mosaic GPU] Add MGPU splash attention implementation

### PR DESCRIPTION
**Pull request by justinjfu**
Adds a splash attention implementation for Mosaic GPU.

The kernel itself is based on the existing FA3 kernel: [attention_mgpu.py](https://github.com/jax-ml/jax/blob/main/jax/experimental/pallas/ops/gpu/attention_mgpu.py)
The masking logic is based on TPU splash attention, but modified heavily to accommodate the GPU kernel: [splash_attention_mask_info.py](https://github.com/jax-ml/jax/blob/main/jax/experimental/pallas/ops/tpu/splash_attention/splash_attention_mask_info.py)

Some sample profiling results using a causal mask. The kernel time include both the time to generate the mask and run the kernel (as measure by the cupti profiler), and in practice the mask time could be amortized over multiple layers. Note that the CUDNN implementation `jax.nn.dot_product_attention` performs very poorly with an explicitly passed-in mask. The underlying kernel supports generating a mask from IDs inside the kernel without materializing the whole mask, but this isn't exposed in JAX.

Forwards pass:
```
==== batch_size=1      kv_seq_len=4096   q_seq_len=4096  num_q_heads=16   head_dim=128    masked=True ====
CUDNN REFERENCE:  435.6  us = 15.8% TC utilization
block_q=64  block_kv=128 max_concurrent_steps=2   :  169.4  us = 40.6% TC utilization
block_q=64  block_kv=64  max_concurrent_steps=2   :  199.7  us = 34.4% TC utilization
block_q=64  block_kv=64  max_concurrent_steps=4   :  202.0  us = 34.0% TC utilization
Best: 169.4  us = 40.6% TC utilization
==== batch_size=1      kv_seq_len=8192   q_seq_len=8192  num_q_heads=16   head_dim=128    masked=True ====
CUDNN REFERENCE:  1723.2 us = 16.0% TC utilization
block_q=64  block_kv=128 max_concurrent_steps=2   :  519.9  us = 52.9% TC utilization
block_q=64  block_kv=64  max_concurrent_steps=2   :  653.0  us = 42.1% TC utilization
block_q=64  block_kv=64  max_concurrent_steps=4   :  658.4  us = 41.8% TC utilization
Best: 519.9  us = 52.9% TC utilization
```

Backwards pass:
```
==== batch_size=1      kv_seq_len=4096   q_seq_len=4096  num_q_heads=16   head_dim=128    masked=True ====
CUDNN REFERENCE:  1548.4 us = 11.1% TC utilization
block_q_dkv=64  block_q_dq=64  block_kv_dq=64  block_kv_dkv=64  max_concurrent_steps=2   :  567.3  us = 30.3% TC utilization
Best: 567.3  us = 30.3% TC utilization
==== batch_size=1      kv_seq_len=8192   q_seq_len=8192  num_q_heads=16   head_dim=128    masked=True ====
CUDNN REFERENCE:  6044.2 us = 11.4% TC utilization
block_q_dkv=64  block_q_dq=64  block_kv_dq=64  block_kv_dkv=64  max_concurrent_steps=2   :  1821.7 us = 37.7% TC utilization
```

### CODE REVIEW

File: jax/experimental/pallas/ops/gpu/splash_attention/hopper_splash_attention.py [hunk 1]
```diff
@@ -0,0 +1,1116 @@
...
+    masked_attention = mask_info is not None
+    if q.ndim != 4 or k.ndim != 4 or v.ndim != 4:
+        raise ValueError(f"q, k, and v should all be 4D, got: {q.ndim=}, {k.ndim=}, {v.ndim=}")
+    batch_size, q_seq_len, num_q_heads, head_dim = q.shape
+    _, kv_seq_len, num_kv_heads, _ = k.shape
+    kv_shape = (batch_size, kv_seq_len, num_kv_heads, head_dim)
+    if k.shape != kv_shape:
+        raise ValueError(f"Expected {k.shape=} to be {kv_shape} (inferred from q)")
+    if k.shape != kv_shape:
+        raise ValueError(f"Expected {v.shape=} to be {kv_shape} (inferred from q)")
```
[the remark this answers is not in the document]
Comment (justinjfu) [reply jax/experimental/pallas/ops/gpu/splash_attention/hopper_splash_attention.py]: good bot

File: jax/experimental/pallas/ops/gpu/splash_attention/attention_mask.py [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+    q_blocks_count, q_mod = divmod(q_seq_len, q_block_size)
+    kv_blocks_count, kv_mod = divmod(kv_seq_len, kv_block_size)
+
+    if q_mod != 0:
+        raise ValueError(f"{q_block_size=} should divide {q_seq_len=}.")
+    if kv_mod != 0:
+        raise ValueError(f"{kv_block_size=} should divide {kv_seq_len=}.")
+
+    if head_count != 1:
+        raise ValueError(f"{head_count=} must be 1.")
```
[the remark this answers is not in the document]
Comment (justinjfu) [reply jax/experimental/pallas/ops/gpu/splash_attention/attention_mask.py]: good bot

File: jax/experimental/pallas/ops/gpu/splash_attention/attention_mask.py [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+    data_next = jnp.broadcast_to(data_next, mask_info_slice_shape)
+    data_next = jnp.where(block_mask == 0, -1, data_next)
+
+    # Compress and tile mask data.
+    # The mask_info is structured such that data blocks are shared between compute warpgroups.
+    # This is due to the structure of a warp-specialized kernel where there is a single memory
+    # warpgroup loading data that can be shared between multiple compute warpgroups, reducing
+    # the overall amount of data copied from HBM.
+    # We do not share the mask between compute warpgroups since the masking pattern typically
+    # differs between different Q-blocks and therefore cannot be shared.
```
Comment (apaszke) [inline jax/experimental/pallas/ops/gpu/splash_attention/attention_mask.py]: Didn't you just say you do share it 5 lines above?

File: jax/experimental/pallas/ops/gpu/splash_attention/hopper_splash_attention.py [hunk 4]
```diff
@@ -0,0 +1,1020 @@
...
+    if barriers:
+        # TODO(justinjfu): Only issue and wait the mask TMA if `block_mask_value=1`
+        for wg in range(len(barriers)):
+
+            @pl.when(lax.axis_index(wg_key) == wg)
+            def _() -> None:
+                plgpu.barrier_wait(barriers[wg])
+
+    def apply_smem_mask(qk, num_nonzeros):
+        mask = plgpu.load(mask_smem_ref, (), layout=plgpu.Layout.WGMMA)
```
Comment (apaszke) [inline jax/experimental/pallas/ops/gpu/splash_attention/hopper_splash_attention.py]: 4 spaces? really? ;)

**Review by apaszke — APPROVED** [review]
This is way too big for me to review (unfortunately), but I gave it a skim and it all looks reasonable (and IIUC is mostly a mashup of TPU splash + the Hopper kernel)

File: jax/experimental/pallas/ops/gpu/splash_attention/attention_mask.py (hunk 3 above)
Comment (justinjfu) [reply jax/experimental/pallas/ops/gpu/splash_attention/attention_mask.py]: No - the data block is shared but the mask is not.

File: jax/experimental/pallas/ops/gpu/splash_attention/hopper_splash_attention.py (hunk 4 above)
Comment (justinjfu) [reply jax/experimental/pallas/ops/gpu/splash_attention/hopper_splash_attention.py]: Oops - fixed to be google style :)
